File: collector_server.ts
import { serve } from "https://deno.land/std@0.203.0/http/server.ts";

const RESULT_FILE = 'resultat.csv';
const IP_FILE = 'ip.txt';
const CONFIG_FILE = 'config.txt';
const HEADER = 'mention;Periode;Type Mention;ID;Timestamp;OfferId\n';
const PORT = Number(Deno.env.get('PORT') || 8000);

const CORS_HEADERS = {
  'Access-Control-Allow-Origin': '*',
  'Access-Control-Allow-Methods': 'GET,POST,OPTIONS',
  'Access-Control-Allow-Headers': 'Content-Type',
};

function jsonResponse(payload: unknown, status = 200) {
  return new Response(JSON.stringify(payload), {
    status,
    headers: { 'Content-Type': 'application/json', ...CORS_HEADERS },
  });
}

function textResponse(message: string, status = 200) {
  return new Response(message, { status, headers: CORS_HEADERS });
}

async function fileExists(path: string) {
  try {
    await Deno.stat(path);
    return true;
  } catch (err) {
    if (err instanceof Deno.errors.NotFound) return false;
    throw err;
  }
}

async function ensureResultFile() {
  if (await fileExists(RESULT_FILE)) {
    const stat = await Deno.stat(RESULT_FILE);
    if (stat.size === 0) {
      await Deno.writeTextFile(RESULT_FILE, HEADER);
    }
    return;
  }
  await Deno.writeTextFile(RESULT_FILE, HEADER);
}

async function readConfig() {
  try {
    const raw = await Deno.readTextFile(CONFIG_FILE);
    return JSON.parse(raw);
  } catch (err) {
    if (err instanceof Deno.errors.NotFound) return {};
    console.warn('Unable to read config', err);
    return {};
  }
}

async function readIpSet() {
  const existing = new Set<string>();
  try {
    const content = await Deno.readTextFile(IP_FILE);
    content.split(/\r?\n/).forEach(line => {
      const ip = line.trim();
      if (ip) existing.add(ip);
    });
  } catch (err) {
    if (!(err instanceof Deno.errors.NotFound)) {
      console.warn('Could not read ip file', err);
    }
  }
  return existing;
}

async function appendCsv(payload: Record<string, string>) {
  await ensureResultFile();
  const line = `${payload.mention || 'non'};${payload.periode || ''};${payload.mentionType || ''};${payload.ip || 'unknown'};${payload.timestamp || ''};${payload.offerId || ''}\n`;
  await Deno.writeTextFile(RESULT_FILE, line, { append: true });
}

async function appendIp(ip: string, existing: Set<string>) {
  if (!ip || ip === 'unknown' || existing.has(ip)) return;
  await Deno.writeTextFile(IP_FILE, ip + '\n', { append: true });
  existing.add(ip);
}

async function handleSubmit(req: Request) {
  let payload: Record<string, string>;
  try {
    payload = await req.json() as Record<string, string>;
  } catch (err) {
    return jsonResponse({ accepted: false, message: 'invalid json' }, 400);
  }
  const config = await readConfig();
  const blockAfterOne = Boolean(config.block_after_1);
  const ip = (payload.ip || 'unknown').toString();

  const existing = await readIpSet();
  if (blockAfterOne && ip !== 'unknown' && existing.has(ip)) {
    return jsonResponse({ accepted: false, message: 'blocked' }, 403);
  }

  await appendCsv(payload);
  await appendIp(ip, existing);
  return jsonResponse({ accepted: true });
}

console.log(`Collector server listening on http://0.0.0.0:${PORT}`);
serve((req) => {
  const url = new URL(req.url);
  if (req.method === 'OPTIONS') {
    return textResponse('', 200);
  }

  if (url.pathname === '/submit' && req.method === 'POST') {
    return handleSubmit(req);
  }

  if (url.pathname === '/ips' && req.method === 'GET') {
    return readIpSet().then(list => jsonResponse([...list]));
  }

  if (url.pathname === '/has_responded' && req.method === 'GET') {
    const ip = url.searchParams.get('ip') || '';
    if (!ip) return jsonResponse({ responded: false });
    return readIpSet().then(list => jsonResponse({ responded: list.includes(ip) }));
  }

  return textResponse('Not found', 404);
}, { port: PORT });

console.log(`Collector server listening on http://0.0.0.0:${PORT}`);

File: collector_server.py
#!/usr/bin/env python3
"""
Simple collector server to receive experiment submissions.
Listens on 0.0.0.0:8000 by default. Endpoints:
- POST /submit  -> JSON payload {ip, offerId, periode, mention, mentionType, timestamp}
- GET /ips     -> returns JSON array of known IPs
- GET /has_responded?ip=1.2.3.4 -> returns JSON {responded: true/false}

It appends results to `resultat.txt` (semicolon-separated) and unique IPs to `ip.txt`.
"""

import http.server
import socketserver
import json
import urllib.parse
import os

PORT = 8000
RESULT_FILE = 'resultat.txt'
IP_FILE = 'ip.txt'
CONFIG_FILE = 'config.txt'

class Handler(http.server.SimpleHTTPRequestHandler):
    def _set_json(self, code=200):
        self.send_response(code)
        self.send_header('Content-Type','application/json')
        self.send_header('Access-Control-Allow-Origin','*')
        self.send_header('Access-Control-Allow-Methods','GET,POST,OPTIONS')
        self.send_header('Access-Control-Allow-Headers','Content-Type')
        self.end_headers()

    def do_OPTIONS(self):
        self._set_json(200)
        self.wfile.write(b'{}')

    def do_GET(self):
        parsed = urllib.parse.urlparse(self.path)
        if parsed.path == '/ips':
            ips = []
            if os.path.exists(IP_FILE):
                with open(IP_FILE,'r',encoding='utf-8') as f:
                    ips = [l.strip() for l in f if l.strip()]
            self._set_json(200)
            self.wfile.write(json.dumps(ips).encode())
            return
        if parsed.path == '/has_responded':
            q = urllib.parse.parse_qs(parsed.query)
            ip = q.get('ip',[''])[0]
            found = False
            if ip and os.path.exists(IP_FILE):
                with open(IP_FILE,'r',encoding='utf-8') as f:
                    found = any(ip == l.strip() for l in f)
            self._set_json(200)
            self.wfile.write(json.dumps({'responded': found}).encode())
            return
        return http.server.SimpleHTTPRequestHandler.do_GET(self)

    def do_POST(self):
        parsed = urllib.parse.urlparse(self.path)
        if parsed.path != '/submit':
            self._set_json(404)
            self.wfile.write(json.dumps({'error':'not found'}).encode())
            return
        length = int(self.headers.get('Content-Length',0))
        raw = self.rfile.read(length).decode('utf-8')
        try:
            data = json.loads(raw)
        except Exception as e:
            self._set_json(400)
            self.wfile.write(json.dumps({'accepted':False,'message':'bad json'}).encode())
            return

        # read server config to know if block_after_1 is enabled
        block_after = False
        try:
            if os.path.exists(CONFIG_FILE):
                with open(CONFIG_FILE,'r',encoding='utf-8') as f:
                    cfg = json.load(f)
                    block_after = bool(cfg.get('block_after_1', False))
        except Exception:
            block_after = False

        ip = data.get('ip','unknown')
        # Check block
        if block_after and os.path.exists(IP_FILE):
            with open(IP_FILE,'r',encoding='utf-8') as f:
                for l in f:
                    if l.strip() == ip:
                        self._set_json(403)
                        self.wfile.write(json.dumps({'accepted':False,'message':'blocked'}).encode())
                        return
        # Append result
        mention = data.get('mention','non')
        periode = data.get('periode','')
        mentionType = data.get('mentionType','')
        timestamp = data.get('timestamp','')
        offerId = data.get('offerId','')
        line = f"{mention};{periode};{mentionType};{ip};{timestamp};{offerId}\n"
        try:
            with open(RESULT_FILE,'a',encoding='utf-8') as f:
                # if file empty, write header
                if os.path.getsize(RESULT_FILE) == 0:
                    f.write('mention;Periode;Type Mention;ID;Timestamp;OfferId\n')
                f.write(line)
            # update ip list
            if ip and ip != 'unknown':
                existing = set()
                if os.path.exists(IP_FILE):
                    with open(IP_FILE,'r',encoding='utf-8') as f:
                        existing = set(l.strip() for l in f if l.strip())
                if ip not in existing:
                    with open(IP_FILE,'a',encoding='utf-8') as f:
                        f.write(ip + '\n')
            self._set_json(200)
            self.wfile.write(json.dumps({'accepted':True}).encode())
        except Exception as e:
            self._set_json(500)
            self.wfile.write(json.dumps({'accepted':False,'message':str(e)}).encode())

if __name__ == '__main__':
    print(f"Starting collector server on port {PORT} (http://0.0.0.0:{PORT})")
    with socketserver.ThreadingTCPServer(('', PORT), Handler) as httpd:
        try:
            httpd.serve_forever()
        except KeyboardInterrupt:
            print('Shutting down')
            httpd.shutdown()
